### PR TITLE
Snow 1105689 git copy

### PR DESCRIPTION
**Pull request by sfc-gh-pczajka**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
* Add `git copy` command
* Refactor `GitManager` class to inherit after `StageManager`

### CODE REVIEW

File: tests/git/test_git_commands.py [hunk 1]
```diff
@@ -53,3 +54,30 @@ def test_fetch(mock_connector, runner, mock_ctx):
 
     assert result.exit_code == 0, result.output
     assert ctx.get_query() == "alter git repository repo_name fetch"
+
+
+@mock.patch("snowflake.connector.connect")
+def test_copy(mock_connector, runner, mock_ctx):
```
Comment (sfc-gh-turbaszek) [inline tests/git/test_git_commands.py:60]: Should we have integration tests?

Comment (sfc-gh-pczajka) [reply tests/git/test_git_commands.py:60]: I'm on it in another PR

Comment (sfc-gh-pczajka) [reply tests/git/test_git_commands.py:60]: https://github.com/snowflakedb/snowflake-cli/pull/865

File: src/snowflake/cli/plugins/git/commands.py [hunk 2]
```diff
@@ -64,3 +67,32 @@ def fetch(
...
+
+
+@app.command(
+    "copy",
+    help="Copies all files from given state of repository to local directory or stage.",
+    requires_connection=True,
+)
+def copy(
+    repository_path: str = RepoPathArgument,
+    destination_path: str = typer.Argument(
```
Comment (sfc-gh-pjob) [inline src/snowflake/cli/plugins/git/commands.py:79]: Can target be a single file? Are we sure it doesn't have to be a directory?

File: src/snowflake/cli/plugins/git/commands.py [hunk 3]
```diff
@@ -92,6 +92,8 @@ def copy(
         )
     else:
         target = Path(destination_path).resolve()
+        if not target.exists():
+            target.mkdir()
```
Comment (sfc-gh-pjob) [inline src/snowflake/cli/plugins/git/commands.py:96]: Shouldn't it be created using secure path?

File: src/snowflake/cli/plugins/git/commands.py [hunk 4]
```diff
@@ -77,7 +77,7 @@ def fetch(
 def copy(
     repository_path: str = RepoPathArgument,
     destination_path: str = typer.Argument(
-        help="Target path for copy operation. Should be stage or local file path.",
+        help="Target path for copy operation. Should be stage or a path to a local directory.",
```
Comment (sfc-gh-pjob) [inline src/snowflake/cli/plugins/git/commands.py:80]: `Should be stage or` - stage or a dir on a stage?

Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/git/commands.py:80]: It can be any stage path: https://docs.snowflake.com/en/sql-reference/sql/copy-files

Comment (sfc-gh-pjob) [reply src/snowflake/cli/plugins/git/commands.py:81]: Ok, I have two questions:

_1: Can I copy a single file from a git repository?

_2: If answer for the first question is "yes", then what happens if I use the following command, what is created on a stage (what paths)?:
```
snow git copy @testrepo/branches/main/abc.txt @teststage/abc.txt
```

Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/git/commands.py:81]: 1. I might have overdid some changes:
currently it creates a directory `@teststage/abc.txt/` and finishes with an error (I'll investigate in a sec)

2. "pure" sql command behaves differently, but also in a way which doesn't seem to be intended:
```
▶ snow sql -q "copy files into @XD/file1.txt from @dummy_repo/branches/main/file.txt"                   
copy files into @XD/file1.txt from @dummy_repo/branches/main/file.txt
+-------------------+
| file              |
|-------------------|
| file1.txtfile.txt |
+-------------------+
```

My proposition is to stick with [1] as we don't have influence over `copy files into` behaviour, but explicitely inform that we expect to see a directory

Comment (sfc-gh-pjob) [reply src/snowflake/cli/plugins/git/commands.py:81]: Yes, I think we should inform that target destination is always a directory - no matter if local or on a stage.

Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/git/commands.py]: Fixed. The downside is that now users will see this error, but it is more readable than previous `java index -1` :D
```
▶ snow git copy @dummy_repo/branches/main @test_stage -c qa        

An unexpected exception occurred. Use --debug option to see the traceback. Exception message:

093507 (42601): 01b2c9c0-0000-ced2-0000-0f2d04a469ba: To access files in git repositories, a slash is required after the branch name, tag name, or commit hash.
```

File: src/snowflake/cli/plugins/git/commands.py [hunk 5]
```diff
@@ -64,3 +81,34 @@ def fetch(
...
+):
+    is_copy = is_stage_path(destination_path)
+    if is_copy:
+        cursor = GitManager().copy_files(
+            source_path=repository_path, destination_path=destination_path
+        )
+    else:
+        target = SecurePath(Path(destination_path).resolve())
+        if not target.exists():
+            target.mkdir()
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/git/commands.py:110]: Why not handle this in get?

File: src/snowflake/cli/plugins/object/stage/manager.py [hunk 6]
```diff
@@ -90,6 +96,13 @@ def put(
             )
         return cursor
 
+    def copy_files(self, source_path: str, destination_path: str) -> SnowflakeCursor:
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/plugins/object/stage/manager.py:99]: Should we consider adding support for 
`snow object stage copy @from_aa @to_bb`?

File: tests/git/test_git_commands.py [hunk 7]
```diff
@@ -53,3 +60,48 @@ def test_fetch(mock_connector, runner, mock_ctx):
...
+    )
+
+
+def test_copy_not_a_stage_error(runner):
+    result = runner.invoke(["git", "copy", "repo_name", "@stage_path/dir_in_stage"])
+    assert result.exit_code == 1
+    _assert_error_message(result.output)
+
+
+def _assert_error_message(output):
```
Comment (sfc-gh-pjob) [inline tests/git/test_git_commands.py:102]:
```suggestion
def _assert_invalid_repo_path_error_message(output):
```

File: src/snowflake/cli/plugins/git/commands.py (hunk 4 above)
Comment (sfc-gh-pjob) [reply src/snowflake/cli/plugins/git/commands.py]: Is it possible now to do the following (copy a single file)?:
```
snow git copy @dummy_repo/branches/main/abc.txt @test_stage -c qa
```

File: src/snowflake/cli/plugins/git/commands.py [hunk 8]
```diff
@@ -13,7 +17,22 @@
 )
 log = logging.getLogger(__name__)
 
+
+def _repo_path_argument_callback(path):
+    if not is_stage_path(path):
+        raise ClickException(
+            f"REPOSITORY_PATH should be a path to git repository stage with scope provided."
+            " For example: @my_repo/branches/main"
```
Comment (sfc-gh-pjob) [inline src/snowflake/cli/plugins/git/commands.py:25]: `@my_repo/branches/main` - is it still a valid example (no `/` at the end)? If it isn't, then please fix it here and in other places. And maybe `RepoPathArgument` help should say when to use `/` at the end and when not.

File: src/snowflake/cli/plugins/object/stage/manager.py (hunk 6 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/object/stage/manager.py:99]: Done :)
https://snowflakecomputing.atlassian.net/browse/SNOW-1213448

File: src/snowflake/cli/plugins/git/commands.py (hunk 4 above)
Comment (sfc-gh-pczajka) [reply src/snowflake/cli/plugins/git/commands.py]: Yes
```
▶ snow git copy @dummy_repo/branches/main/file.txt @test_stage
copy files into @XD2/ from @dummy_repo/branches/main/file.txt
+----------+
| file     |
|----------|
| file.txt |
+----------+

▶ snow object stage list test_stage
ls @test_stage
+----------------------------------------------------------------------------------------------+
| name                | size | md5                              | last_modified                |
|---------------------+------+----------------------------------+------------------------------|
| test_stage/file.txt | 16   | e0cd340fc9125e498642fcf7cd2463dc | Wed, 6 Mar 2024 12:03:06 GMT |
+----------------------------------------------------------------------------------------------+

```

File: tests/git/test_git_commands.py (hunk 7 above)
Comment (sfc-gh-pjob) [reply tests/git/test_git_commands.py:102]: I don't see any change about it :)

File: tests/git/test_git_commands.py [hunk 9]
```diff
@@ -53,3 +60,48 @@ def test_fetch(mock_connector, runner, mock_ctx):
...
+    assert "Error" in output
+    assert (
+        "REPOSITORY_PATH should be a path to git repository stage with scope" in output
+    )
+    assert "provided. For example: @my_repo/branches/main" in output
```
Comment (sfc-gh-pjob) [inline tests/git/test_git_commands.py:107]:
```suggestion
    assert "provided. For example: @my_repo/branches/main/" in output
```

**Review by sfc-gh-pjob — APPROVED** [review]

File: src/snowflake/cli/api/secure_path.py [hunk 10]
```diff
@@ -271,20 +271,20 @@ def temporary_directory(cls):
             log.info("Removing temporary directory %s", tmpdir)
 
     def _assert_exists_and_is_file(self) -> None:
-        self._assert_exists()
-        self._assert_is_file()
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/secure_path.py:275]: Separate to other change to as this change seems to be unrelated.

Comment (sfc-gh-pczajka) [reply src/snowflake/cli/api/secure_path.py:275]: Done: https://github.com/snowflakedb/snowflake-cli/pull/875
